node_modules/col-activities/lib/notifications/daily.js

```diff
@@ -0,0 +1,548 @@
+ * @param  {User}     currentUser   The current user
+ * @return {String}                 A string representation of the set of users
+ * @api private
+ */
+var getSummaryActors = function(users, currentUser) {
+  var actors = _.chain(users)
+    .uniq('id')
+    .filter(function(user) {
+      return (user.id !== currentUser.id);
+    })
+    .value()
+
+  if (actors.length === 1) {
+    return actors[0].canvas_full_name
+  } else if (actors.length === 2) {
+    return actors[0].canvas_full_name + ' and ' + actors[1].canvas_full_name
+  } else if (actors.length === 3) {
+    return actors[0].canvas_full_name + ', ' + actors[1].canvas_full_name + ' and 1 other'
+  } else {
+    var nOthers = actors.length - 2;
+    return actors[0].canvas_full_name + ', ' + actors[1].canvas_full_name + ' and ' + nOthers + ' others';
+  }
+};
+
+
+/* Data retrieval */
+
+
+/**
+ * Get the course data that had any activity in the last day. This is limited to:
+ *  - assets that were commented on
+ *  - assets where a comment was placed on a reply
+ *  - whiteboards that received chat messages
+ *
+ * @param  {Context}    course                        The course for which to get the course data
+ * @param  {Function}   callback                      Standard callback
+ * @param  {Object}     callback.err                  An error object, if any
+ * @param  {Object}     callback.data                 The relevant daily data in the course
+ * @param  {Object}     callback.data.users           All the users in the course (including those that weren't active)
+ * @param  {Object}     callback.data.assets          Assets with their recent comments and replies
+ * @param  {Object}     callback.data.whiteboards     Whiteboards with their recent chat messages
+ * @api private
+ */
+var getCourseData = function(course, callback) {
+  var ctx = {
+    'course': course,
+
+    // Fake an admin user. At this point we can't reliably retrieve an admin from
+    // the database as there might not be any yet
+    'user': {
+      'is_admin': true
+    }
+  };
+
+  // 1. Get all the users in this course. We need to retrieve all users as we might need to display
+  // a user's name in the email even if they weren't active. (For example, the parent comment of a
+  // recent reply)
+  var enrollmentStates = _.values(CollabosphereConstants.ENROLLMENT_STATE);
+  UsersAPI.getAllUsers(ctx, enrollmentStates, function(err, users) {
+    if (err) {
+      log.error({'err': err, 'course': ctx.course.id}, 'Failed to retrieve the users for a course');
+      return callback({'code': 500, 'msg': 'Failed to retrieve the users for a course'});
+    }
+
+    // Index the users by their id. This allows for quicker lookups later on
+    users = _.chain(users)
+      .map(function(user) {
+        return user.toJSON();
+      })
+      .indexBy('id')
+      .value();
+
+    // We need all the data since yesterday
+    var yesterday = moment().subtract(1, 'day').format('YYYY-MM-DD HH:mm:ss')
```

Comment:
I think this has the potential of including certain activities in 2 emails and not including certain activities at all, especially once we have a large number of courses:

- An activity took place at 2 minutes past the hour, but this code is executed 5 minutes past the hour. The next day, this code is executed 1 minute after the hour. This means that it would be included into the first email and the second email
- This code is executed at 2 minutes after the hour and an activity takes place at 3 minutes after the hour. The activity is not included in the email. The next day, this code is executed at 4 minutes after the hour and the activity is again not included in the email.

We should probably base the yesterday and today timestamps on the actual configured email time.